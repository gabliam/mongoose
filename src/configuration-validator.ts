import * as Joi from 'joi';

export const configurationValidator = Joi.object()
  .keys({
    uri: Joi.string(),
    host: Joi.string(),
    database_name: Joi.string(),
    port: Joi.number().default(27017),
    options: Joi.object()
  })
  .xor('uri', 'host')
  .and('host', 'database_name');
