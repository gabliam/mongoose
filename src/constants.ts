/** Type in registry */
export const TYPE = {
  Document: 'Document',
};

export const METADATA_KEY = {
  document: '_document'
};

export const LIST_DOCUMENT = Symbol('LIST_DOCUMENT');
