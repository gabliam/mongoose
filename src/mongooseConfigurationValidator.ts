import * as Joi from 'joi';

export const mongooseConfigurationValidator = Joi.object().keys({
  uri: Joi.string(),
  host: Joi.string(),
  database_name: Joi.string(),
  port: Joi.number(),
  options: Joi.object(),
}).xor('uri', 'host').and('host', 'database_name');
