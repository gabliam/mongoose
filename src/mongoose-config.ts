import { PluginConfig, Value, Bean, inject } from '@gabliam/core';
import { LIST_DOCUMENT } from './constants';
import { configurationValidator } from './configuration-validator';
import { MongooseConfiguration } from './interfaces';
import { MongooseConnection } from './mongoose-connection';
import * as d from 'debug';

const debug = d('Gabliam:Plugin:mongoose');

@PluginConfig()
export class PluginMongooseConfig {

  @Value('application.mongoose', configurationValidator)
  mongooseConfiguration: MongooseConfiguration;

  listDocument: any[];

  constructor(
    @inject(LIST_DOCUMENT) listDocument: any[]
  ) {
    debug('constructor PluginMongooseConfig');
    this.listDocument = listDocument;
  }

  @Bean(MongooseConnection)
  create() {
    debug('Create MongooseConnection', this.mongooseConfiguration);
    return new MongooseConnection(this.mongooseConfiguration, this.listDocument);
  }
}
