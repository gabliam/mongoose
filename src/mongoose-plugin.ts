import {
  interfaces as coreInterfaces,
  inversifyInterfaces,
  Scan,
  Registry,
  Plugin
} from '@gabliam/core';
import { TYPE, LIST_DOCUMENT } from './constants';
import * as d from 'debug';
const debug = d('Gabliam:Plugin:mongoose');
import { MongooseConnection } from './mongoose-connection';

@Plugin()
@Scan()
export class MongoosePlugin implements coreInterfaces.GabliamPlugin {
  bind(container: inversifyInterfaces.Container, registry: Registry) {
    const documents = registry.get(TYPE.Document).map(({ id, target }) => {
      return target;
    });
    debug('list documents', documents);
    container.bind<any>(LIST_DOCUMENT).toConstantValue(documents);
  }

  async destroy(container: inversifyInterfaces.Container, registry: Registry) {
    const connection = container.get<MongooseConnection>(MongooseConnection);
    if (connection) {
      await connection.conn.close();
    }
  }
}
