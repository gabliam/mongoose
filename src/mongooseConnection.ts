import * as mongoose from 'mongoose';
import { Repository } from './repository';
import { METADATA_KEY } from './constants';
import { DocumentMetadata } from './interfaces';
const { Mongoose } = mongoose;

(<any>mongoose).Promise = global.Promise;

export class MongooseConnection {
    public conn: mongoose.Connection;

    private repositories = new Map<string, Repository<any>>();

    private schemas = new Map<string, mongoose.Model<any>>();

    constructor(uri: string, options: mongoose.ConnectionOptions, listDocument: any[]) {
        let m = new Mongoose();
        this.conn = m.createConnection(uri, options);
        for (let document of listDocument) {
            let { collectionName, name, schema } = <DocumentMetadata>Reflect.getOwnMetadata(METADATA_KEY.document, document);
            let documentName = name.toLowerCase();
            let clazzSchema = this.conn.model<mongoose.Document>(name, schema, collectionName);
            this.schemas.set(documentName, clazzSchema);
        }
    }

    getRepository<T>(documentName: string): Repository<T>;
    getRepository<T>(clazz: any): Repository<T> {
        if (typeof clazz === 'string') {
            return this.getRepositoryByName<T>(clazz);
        }

        let { collectionName, name, schema } = <DocumentMetadata>Reflect.getOwnMetadata(METADATA_KEY.document, clazz);
        let documentName = name.toLowerCase();
        if (this.repositories.has(documentName)) {
            return this.repositories.get(documentName);
        }

        let clazzSchema = this.conn.model<T & mongoose.Document>(name, schema, collectionName);
        let repository = new Repository<T>(clazzSchema);
        this.repositories.set(documentName, repository);
        return repository;
    }

    private getRepositoryByName<T>(documentName: string) {
        let name = documentName.toLowerCase();
        if (this.repositories.has(name)) {
            return this.repositories.get(name);
        } else {
            if (this.schemas.has(name)) {
                let repository = new Repository<T>(this.schemas.get(name));
                this.repositories.set(name, repository);
                return repository;
            }
            throw new Error(`Unknown repository ${name}`);
        }
    }
}
