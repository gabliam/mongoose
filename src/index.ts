import * as interfaces from './interfaces';
import { MongoosePlugin } from './mongoose-plugin';

export * from './decorators';
export { interfaces };
export * from './mongoose-connection';
export * from './repository';
export { MongoosePlugin as default };
