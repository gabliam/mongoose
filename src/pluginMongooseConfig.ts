import { PluginConfig, Value, optional, Bean, inject } from '@gabliam/core';
import { MongooseConnectionOptionsBeanId, LIST_DOCUMENT } from './constants';
import { MongooseConnection } from './mongooseConnection';
import * as mongoose from 'mongoose';
import * as d from 'debug';

const debug = d('Gabliam:Plugin:mongoose');

@PluginConfig()
export class PluginMongooseConfig {

    @Value('application.mongoose.uri')
    uri: string;

    @Value('application.mongoose.host')
    host: string;

    @Value('application.mongoose.database_name')
    database_name: string;

    @Value('application.mongoose.port')
    port: string;

    @Value('application.mongoose.options')
    connectionOptions: mongoose.ConnectionOptions;

    listDocument: any[];

    constructor(
        @inject(MongooseConnectionOptionsBeanId) @optional() connectionOptions: mongoose.ConnectionOptions,
        @inject(LIST_DOCUMENT) listDocument
    ) {
        debug('constructor PluginMongooseConfig', connectionOptions);
        this.connectionOptions = connectionOptions;
        this.listDocument = listDocument;
    }

    @Bean('MongooseConnection')
    create() {
        let mongooseConnection = new MongooseConnection(this.uri, this.connectionOptions, this.listDocument);
        return mongooseConnection;
    }
}
