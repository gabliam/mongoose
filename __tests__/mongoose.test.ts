import { MongoosePluginTest } from './mongoose-plugin-test';
import { Document, MongooseConnection } from '../src';
import { Gabliam } from '@gabliam/core';
import * as mongoose from 'mongoose';

interface HeroModel {
  name: string;
}

let appTest: MongoosePluginTest;
beforeEach(async () => {
  appTest = new MongoosePluginTest();
});

afterEach(async () => {
  try {
    await appTest.destroy();
  } catch (e) {}
});

describe('Errors', () => {
  test('without config folder', async () => {
    await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
  });

  test('with bad config', async () => {
    appTest.addConf('application.mongoose', 'bad');
    await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
  });

  test('with host without database_name', async () => {
    appTest.addConf('application.mongoose', {
      host: 'localhost'
    });
    await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
  });

  test('with host and database_name and unknown value', async () => {
    appTest.addConf('application.mongoose', {
      host: 'localhost',
      database_name: 'test',
      lol: 'test'
    });
    await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
  });
  test('bad repo', async () => {
    appTest.addConf(
      'application.mongoose.uri',
      'mongodb://127.0.0.1/mongoosetest'
    );
    await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);
    const connection = appTest.gab.container.get(MongooseConnection);
    expect(() => connection.getRepository('HeroBad')).toThrowError();
  });
});

test('with config uri', async () => {
  appTest.addConf(
    'application.mongoose.uri',
    'mongodb://127.0.0.1/mongoosetest'
  );

  @Document('Hero')
  class Hero {
    static getSchema() {
      return new mongoose.Schema({
        name: {
          type: String,
          required: true
        }
      });
    }
  }

  appTest.addClass(Hero);

  await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);
  const connection = appTest.gab.container.get(MongooseConnection);
  const repo = connection.getRepository<HeroModel>('Hero');
  let res = await repo.findAll();
  expect(res).toMatchSnapshot();
  await repo.create({
    name: 'test'
  });
  res = await repo.findAll();
  expect(res[0].name).toMatchSnapshot();
  await connection.conn.dropDatabase();
});

test('with config host & database', async () => {
  appTest.addConf('application.mongoose', {
    host: '127.0.0.1',
    database_name: 'mongoosetest'
  });

  @Document('Hero')
  class Hero {
    static getSchema() {
      return new mongoose.Schema({
        name: {
          type: String,
          required: true
        }
      });
    }
  }

  appTest.addClass(Hero);

  await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);
  const connection = appTest.gab.container.get(MongooseConnection);
  const repo = connection.getRepository<HeroModel>('Hero');
  let res = await repo.findAll();
  expect(res).toMatchSnapshot();
  await repo.create({
    name: 'test'
  });
  res = await repo.findAll();
  expect(res[0].name).toMatchSnapshot();
  await connection.conn.dropDatabase();
});
