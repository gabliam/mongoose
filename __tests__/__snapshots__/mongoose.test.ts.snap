// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Errors with bad config 1`] = `[Error: Error for 'application.mongoose' value{"isJoi":true,"name":"ValidationError","details":[{"message":"\\"value\\" must be an object","path":"value","type":"object.base","context":{"key":"value"}}],"_object":"bad"}]`;

exports[`Errors with host and database_name and unknown value 1`] = `[Error: Error for 'application.mongoose' value{"isJoi":true,"name":"ValidationError","details":[{"message":"\\"lol\\" is not allowed","path":"lol","type":"object.allowUnknown","context":{"child":"lol","key":"lol"}}],"_object":{"host":"localhost","database_name":"test","lol":"test"}}]`;

exports[`Errors with host without database_name 1`] = `[Error: Error for 'application.mongoose' value{"isJoi":true,"name":"ValidationError","details":[{"message":"\\"value\\" contains [host] without its required peers [database_name]","path":"value","type":"object.and","context":{"present":["host"],"presentWithLabels":["host"],"missing":["database_name"],"missingWithLabels":["database_name"],"key":"value"}}],"_object":{"host":"localhost"}}]`;

exports[`Errors without config folder 1`] = `[TypeError: Cannot read property 'host' of undefined]`;

exports[`with config host & database 1`] = `Array []`;

exports[`with config host & database 2`] = `"test"`;

exports[`with config uri 1`] = `Array []`;

exports[`with config uri 2`] = `"test"`;
